pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1211,40 +1226,65 @@ func getUlAddrs(ctx *zedrouterContext,
 		status.BridgeMac.String())
 	var err error
 	var addr string
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
-	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
-	} else {
-		bridgeIPAddr = addr
+	var appNumOnUNet int
+
+	// App Static IP Address matches with Bridge IP Address
```

Comment:
Adding a check that a static IP doesn't conflict with a dynamic IP makes a lot of sense.
But shouldn't that apply to all IP addresses and not just the bridge IP address?
And should we record the static IP allocations to make sure they are not used for dynamic allocation later?
If we do the last thing it means that it will have some impact on the use of appNumOnUNet usage. Without it, we could just use dhcpRange.Start +  appNumOnUNet, but if the static allocations need to punch holes in the range we need some more logic in the IP allocation. Something we shoud talk about.

Reply:
> I think this is lacking structure and layering, making it very hard to understand and there might be bugs in here.
> 
> I think the IP address allocation needs to be completely separated from the management and allocation and persistency of the appNumOnUNet. Then the IP address allocation will just retrieve the appNumOnUNet and pick that sequence number in the range (perhaps with appNumOnUNet zero being reserved for the bridgeIPAddr.
> 
> My point is that the conflation of the appNumOnUNet and IP address allocation makes this far far too complex to understand.
> 
> The appNumOnUNet management code should be at least in a separate file if not in a separate package with well defined APIs.
> 
> That API should presumably just have calls from when an network instance is created and deleted and a call when an app instance is created and deleted. AFAIU what they need are:
> Network instance create: reserve appNumOnUNet zero for any later use for the bridgeIPAddr on that network instance UUID.
> Network instance delete: delete all records for that network instance
> App instance create: allocate the lowest unused appNumOnUNet for that network instance UUID. Question: Should this handle the case of an app instance having multiple adapters to the same network instance? We don't prohibit that anywhere in the code.
> App instance delete: delete all record of the app instance's allocations.
> TBD: Do we need to handle a modify to an AppNetworkStatus where it gets a different set of UnderlayNetworkConfig? If so we need to call the function used at App Instance create for any new ulConfigs.
> 
> With that set of logic, the IP allocation code can just retrieve the appNumOnUnet because it already exists, and check that its value fits in the range, and do the same recordIPAssignment as before.
> 
> Can we iterate on a daily basis on the API description and comments for that new file/package next week?

Sure, will do some more cleanup.